From 76728893475a94b4f23aa58bae3b8c56cae62585 Mon Sep 17 00:00:00 2001
From: Suvidh Subodh Hegde <suvidhsu@qti.qualcomm.com>
Date: Fri, 14 Aug 2026 11:00:59 +0530
Subject: [PATCH] Fix for Topology notification not sent during client
 dissociation

Root Cause: After the extender completes CONTROLLER_DISCOVERY,
topology notifications should continue to be reported for both
client association and disassociation events.
While association events were already being handled correctly, disassociation events were not.

Resolution: Added handling for client disassociation events to
ensure topology notifications are generated and reported correctly after CONTROLLER_DISCOVERY completion.

Signed-off-by: Suvidh Subodh Hegde <suvidhsu@qti.qualcomm.com>
---
 agent/src/beerocks/slave/son_slave_thread.cpp | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

diff --git a/agent/src/beerocks/slave/son_slave_thread.cpp b/agent/src/beerocks/slave/son_slave_thread.cpp
index 1df2e01f6..fb3f64f10 100644
--- a/agent/src/beerocks/slave/son_slave_thread.cpp
+++ b/agent/src/beerocks/slave/son_slave_thread.cpp
@@ -4083,7 +4083,19 @@ bool slave_thread::handle_cmdu_ap_manager_message(const std::string &fronthaul_i
         // notify master
         if (!link_to_controller()) {
             LOG(DEBUG) << "Controller is not connected";
-            return true;
+
+            LOG(DEBUG) << "controller mac: " << db->controller_info.bridge_mac
+                       << " agent_bridge_mac: " << db->bridge.mac;
+
+            if ((db->controller_info.bridge_mac != network_utils::ZERO_MAC) &&
+                              (db->controller_info.bridge_mac == db->bridge.mac)) {
+                LOG(DEBUG) << " Local Agent - Send topology notification to controller ";
+            } else if (db->statuses.controller_discovery_completed == true) {
+                LOG(DEBUG) << " Controller discovery completed in Non-Local Agent - send topology notification";
+            } else {
+                LOG(ERROR) << " Controller discovery not completed in Non-Local Agent - Drop topology notification";
+                return true;
+            }
         }
 
         if (send_leave) {
@@ -4510,7 +4522,6 @@ bool slave_thread::handle_cmdu_ap_manager_message(const std::string &fronthaul_i
 
             LOG(DEBUG) << "controller mac: " << db->controller_info.bridge_mac << " agent_bridge_mac: " << db->bridge.mac; 
 
-            //TODO: Needs to handle Extender Agent only mode
             if ((db->controller_info.bridge_mac != network_utils::ZERO_MAC) && 
                               (db->controller_info.bridge_mac == db->bridge.mac)) {
                 LOG(DEBUG) << " Local Agent - Send topology notification to controller " ;
-- 
2.34.1

